# Sixth Exercise Session

In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'

In [16]:
import pandas as pd
import numpy as np
!pip install recordlinkage
import recordlinkage
from recordlinkage.datasets import load_febrl1, load_febrl2

## Data Integration as Data Union

In the following example we are going to integrate two datasets that have the same schema, i.e. we are performing data union task.

In [9]:
data1 = load_febrl1()
data1.head(10)

given_name     surname street_number          address_1  \
rec_id                                                                  
rec-223-org          NaN      waller             6   tullaroop street   
rec-122-org      lachlan       berry            69      giblin street   
rec-373-org       deakin  sondergeld            48  goldfinch circuit   
rec-10-dup-0       kayla  harrington           NaN     maltby circuit   
rec-227-org         luke      purdon            23       ramsay place   
rec-6-dup-0          NaN   trevorrow            16       dumas street   
rec-190-dup-0     darcie      turtur            10     blacket street   
rec-294-org      william      bishop            21      neworra place   
rec-206-dup-0        NaN    lombardi            36      yerralee road   
rec-344-org          NaN      julius            52       florey drive   

                             address_2         suburb postcode state  \
rec_id                                                                 
rec-223-org                   willaroo       st james     4011    wa   
rec-122-org                  killarney        bittern     4814   qld   
rec-373-org                    kooltuo     canterbury     2776   vic   
rec-10-dup-0                   coaling       coolaroo     3465   nsw   
rec-227-org                     mirani        garbutt     2260   vic   
rec-6-dup-0    2/98-latchford barracks  mount immaney     2281    wa   
rec-190-dup-0                   eureka  beverly hills     2263   nsw   
rec-294-org                   apmnt 65      worongary     6225   qld   
rec-206-dup-0      leisure living vlge      carlsruhe     3149   qld   
rec-344-org       north stirling downs       coolaroo     2259   qld   

              date_of_birth soc_sec_id  
rec_id                                  
rec-223-org        19081209    6988048  
rec-122-org        19990219    7364009  
rec-373-org        19600210    2635962  
rec-10-dup-0       19150612    9004242  
rec-227-org        19831024    8099933  
rec-6-dup-0        19530313    4811642  
rec-190-dup-0           NaN    2025650  
rec-294-org        19490130    9773843  
rec-206-dup-0      19870919    1613132  
rec-344-org        19630521    1797144

In [10]:
data2 = load_febrl2()
data2.head(10)

given_name    surname street_number         address_1  \
rec_id                                                                
rec-2778-org       sarah      bruhn            44     forbes street   
rec-712-dup-0      jacob     lanyon             5        milne cove   
rec-1321-org     brinley  efthimiou            35  sturdee crescent   
rec-3004-org     aleisha     hobson            54     oliver street   
rec-1384-org       ethan    gazzola            49    sheaffe street   
rec-3981-org      alicia       hope           100   mansfield place   
rec-916-org     benjamin   kolosche            78     keenan street   
rec-1684-org    petreece     lusted            78      baseby place   
rec-63-dup-0      olivia      white            55      duffy street   
rec-3808-org         NaN    maynard            32     yerralee road   

                      address_2              suburb postcode state  \
rec_id                                                               
rec-2778-org         wintersloe        kellerberrin     4510   vic   
rec-712-dup-0           wellwod  beaconsfield upper     2602   vic   
rec-1321-org          tremearne         scarborough     5211   qld   
rec-3004-org          inglewood           toowoomba     3175   qld   
rec-1384-org         bimby vale          port pirie     3088    sa   
rec-3981-org             sunset              byford     6061    sa   
rec-916-org             wingara     raymond terrace     3212    sa   
rec-1684-org                NaN       east branxton     6214    sa   
rec-63-dup-0   shopping village          mirrabooka     2260   vic   
rec-3808-org                NaN      altona meadows     4019   nsw   

              date_of_birth soc_sec_id  
rec_id                                  
rec-2778-org       19300213    7535316  
rec-712-dup-0      19080712    9497788  
rec-1321-org       19940319    6814956  
rec-3004-org       19290427    5967384  
rec-1384-org       19631225    3832742  
rec-3981-org       19421201    7934773  
rec-916-org        19450918    5698873  
rec-1684-org       19950620    8084076  
rec-63-dup-0       19000106    4996142  
rec-3808-org       19150402    5146525

In [12]:
print(data1.shape)
print(data2.shape)

(1000, 10)
(5000, 10)


In [17]:
indexer = recordlinkage.Index()
#indexer.block(left_on='first_name', right_on='given_name')
indexer.block('given_name')
pairs = indexer.index(data1, data2)
len(pairs)

<Index>

18000

The left_on and right_on parameters let you to link attributes of two different dataset providing their name in the two schemas.

In [18]:
compare_c1 = recordlinkage.Compare()

compare_c1.exact('given_name', 'given_name', label='given_name')
compare_c1.string('surname', 'surname', method='jarowinkler', threshold=0.85, label='surname')
compare_c1.exact('date_of_birth', 'date_of_birth', label='date_of_birth')
compare_c1.exact('suburb', 'suburb', label='suburb')
compare_c1.exact('state', 'state', label='state')
compare_c1.string('address_1', 'address_1', threshold=0.85, label='address_1')

features = compare_c1.compute(pairs, data1, data2)
features.head()

<Compare>

<Compare>

<Compare>

<Compare>

<Compare>

<Compare>

given_name  surname  date_of_birth  suburb  state  \
rec_id_1    rec_id_2                                                           
rec-122-org rec-242-org             1      0.0              0       0      0   
            rec-320-org             1      0.0              0       0      0   
            rec-534-dup-2           1      0.0              0       0      0   
            rec-482-org             1      0.0              0       0      0   
            rec-2718-org            1      0.0              0       0      0   

                           address_1  
rec_id_1    rec_id_2                  
rec-122-org rec-242-org          0.0  
            rec-320-org          0.0  
            rec-534-dup-2        0.0  
            rec-482-org          0.0  
            rec-2718-org         0.0

In [19]:
features.sum(axis=1).value_counts().sort_index(ascending=False)

3.0       14
2.0     4098
1.0    13888
dtype: int64

## Record Linkage in Data Integration

In [21]:
hospital_accounts = pd.read_csv('./Dataset/hospital_account_info.csv')
hospital_reimbursement = pd.read_csv('./Dataset/hospital_reimbursement.csv')

In [22]:
hospital_accounts.head()

Account_Num                      Facility Name                    Address  \
0        10605             SAGE MEMORIAL HOSPITAL  STATE ROUTE 264 SOUTH 191   
1        24250        WOODRIDGE BEHAVIORAL CENTER       600 NORTH 7TH STREET   
2        10341           DOUGLAS GARDENS HOSPITAL            5200 NE 2ND AVE   
3        81101  SUNCOAST BEHAVIORAL HEALTH CENTER             4480 51ST ST W   
4        39835           TREASURE VALLEY HOSPITAL   8800 WEST EMERALD STREET   

           City State  ZIP Code County Name    Phone Number  \
0        GANADO    AZ     86505      APACHE  (928) 755-4541   
1  WEST MEMPHIS    AR     72301  CRITTENDEN  (870) 394-4113   
2         MIAMI    FL     33137  MIAMI-DADE  (305) 751-8626   
3     BRADENTON    FL     34210     MANATEE  (941) 792-2222   
4         BOISE    ID     83704         ADA  (208) 373-5000   

               Hospital Type              Hospital Ownership  
0  Critical Access Hospitals  Voluntary non-profit - Private  
1                Psychiatric                     Proprietary  
2       Acute Care Hospitals  Voluntary non-profit - Private  
3                Psychiatric                     Proprietary  
4       Acute Care Hospitals                     Proprietary

In [24]:
hospital_reimbursement.head()

Provider_Num                     Provider Name     Provider Street Address  \
0        839987  SOUTHEAST ALABAMA MEDICAL CENTER      1108 ROSS CLARK CIRCLE   
1        519118     MARSHALL MEDICAL CENTER SOUTH  2505 U S HIGHWAY 431 NORTH   
2        733073    ELIZA COFFEE MEMORIAL HOSPITAL          205 MARENGO STREET   
3        201752          MIZELL MEMORIAL HOSPITAL               702 N MAIN ST   
4        678488                 ST VINCENT'S EAST  50 MEDICAL PARK EAST DRIVE   

  Provider City Provider State  Provider Zip Code  Total Discharges  \
0        DOTHAN             AL              36301               118   
1          BOAZ             AL              35957                43   
2      FLORENCE             AL              35631                73   
3           OPP             AL              36467                12   
4    BIRMINGHAM             AL              35235                74   

   Average Covered Charges  Average Total Payments  Average Medicare Payments  
0                 20855.61                 5026.19                    4115.52  
1                 13289.09                 5413.63                    4490.93  
2                 22261.60                 4922.18                    4021.79  
3                 10901.33                 5343.50                    4284.17  
4                 28117.95                 5947.12                    4819.53

This is a classical example of **record linkage inside a data integration process**.

We have two dataset refferring to hospitals: one considering some aspect and the other other aspects, but there is a common part. Hence the difference w.r.t. the former example is that here the schemas are not fully overlapping but ony partially, i.e. they carry different information about the same objects.

We will use Sorted Neighborhood method since it works better.

In [25]:
indexer = recordlinkage.SortedNeighbourhoodIndex(left_on = 'Facility Name', right_on='Provider Name', window=15)

Hence the dataset are sorted using the dataset's attributes we are specifing and then the merge is done inside the windows working as the normal windows approach.

In [27]:
pairs = indexer.index(hospital_accounts, hospital_reimbursement)
print(len(pairs))

38900


Now we have to tell the sistem how the matching should be done, i.e. we have to specify the common attributes saying which atrtibute should be compared with each attribute.

In [30]:
compare_c1 = recordlinkage.Compare()

compare_c1.exact('Facility Name', 'Provider Name', label='Hospital')
compare_c1.string('Address', 'Provider Street Address', method='jarowinkler', threshold=0.85, label='Address')
compare_c1.string('City', 'Provider City', threshold=0.85, label='City')
compare_c1.exact('State', 'Provider State', label='State')
compare_c1.exact('ZIP Code', 'Provider Zip Code', label='ZIP Code')

features = compare_c1.compute(pairs, hospital_accounts, hospital_reimbursement)
features.head()

<Compare>

<Compare>

<Compare>

<Compare>

<Compare>

Hospital  Address  City  State  ZIP Code
0 717          0      0.0   0.0      0         0
1 516          0      0.0   0.0      0         0
2 132          0      0.0   0.0      0         0
4 2168         0      0.0   0.0      0         0
6 1430         0      0.0   0.0      0         0

In [31]:
features.sum(axis=1).value_counts().sort_index(ascending=False)

5.0     2240
4.0      129
3.0       48
2.0      280
1.0     5442
0.0    30761
dtype: int64

We can notice there is a quite high similarity sice there are a lot of tuples that has all the 5 attributes matching.

In [36]:
matches = features[features.sum(axis=1) > 3]
print(len(matches))
matches

2240


Hospital  Address  City  State  ZIP Code
14   1777         1      1.0   1.0      1         1
27   1502         1      1.0   1.0      1         1
28   779          1      1.0   1.0      1         1
1293 1115         1      1.0   1.0      1         1
3109 507          1      1.0   1.0      1         1
...             ...      ...   ...    ...       ...
5326 2400         1      1.0   1.0      1         1
5328 65           1      1.0   1.0      1         1
5329 2213         1      1.0   1.0      1         1
5334 988          1      1.0   1.0      1         1
5337 593          1      1.0   1.0      1         1

[2240 rows x 5 columns]

For these 2240 matches we can built an integrated dataset: we can put them together and retrieve the data from the two dataset creating a new richer schema.

**From this record linkage you should perform data fusion to build the integrated schema** and all the operation needed to obtain a good integration.